### DISCUSSION TITLE
Permission management on individual contributions

### DISCUSSION THREAD

**Discussion by morzan1001** [appwrite/appwrite#8112] [Q&A]
Hey, I'm playing around with appwrite to see if it's for me and I want to use it in backends in the future. However, I've run into a problem that you might be able to help me with. 

In a normal blog application, a user should be able to create, delete and edit posts. Other users should be able to see these posts but not edit or delete them. I have created a collection in which the posts can be stored, but now I have to give the users all the permissions for delete and update although each user should only be able to edit his own post. How do I implement this with the appwrite permissions? Does anyone have a suggestion for me? Thank you very much

**Comment by Saakhavat-Prime** [comment]
Appwrite handles this using **document-level permissions**, so you don't need to give every user global update or delete access to the collection.

A common approach for a blog application is:

* Allow authenticated users to create documents in the collection.
* When a user creates a post, set the document permissions like this:

  * **Read:** Everyone (`Role.any()`) or authenticated users (`Role.users()`), depending on your requirements.
  * **Update:** Only the creator (`Role.user(user.$id)`).
  * **Delete:** Only the creator (`Role.user(user.$id)`).

Example:

```javascript
await databases.createDocument(
  DATABASE_ID,
  COLLECTION_ID,
  ID.unique(),
  {
    title: "My First Post",
    content: "Hello World!",
    authorId: user.$id
  },
  [
    Permission.read(Role.any()),
    Permission.update(Role.user(user.$id)),
    Permission.delete(Role.user(user.$id))
  ]
);
```

This way:

* ✅ Any user can read posts (if `Role.any()` is used).
* ✅ Users can edit and delete only their own posts.
* ❌ Users cannot modify posts created by other users.

It's also a good idea to store the `authorId` in the document. This makes it easy to query a user's posts or display the author's information. However, `authorId` alone does **not** provide security—the document permissions are what actually enforce who can update or delete the post.

**Comment by krishsharma-code** [comment]
To add to @Saakhavat-Prime 's excellent code-based approach, if you want to understand how this looks from the **Appwrite Console (Dashboard UI)** side, here is how you can visualize and set it up:

1. Navigate to your **Appwrite Console** > **Databases** > Select your Blog Database > Select your Posts Collection.
2. Go to the **Settings** tab of that Collection.
3. Scroll down to **Permissions** and ensure **Document Security** is **Enabled**. (This is the most crucial step for document-level permissions to work).
4. In the Collection-level permissions, you only need to grant **Read** and **Create** access to `Users` (authenticated users). 
5. Leave **Update** and **Delete** blank at the collection level. 

By keeping Update/Delete blank at the collection level and passing `Permission.update(Role.user(user.$id))` during document creation (as shown in the code above), Appwrite's backend will strictly guarantee that only the original author can modify or delete that specific post. 

Hope this combined with the code snippet clears up the confusion! 🚀
